src/mpid/ch4/netmod/ucx/ucx_progress.c

```diff
@@ -78,44 +75,34 @@ static int am_handler(void *msg, size_t msg_sz)
     if (target_cmpl_cb) {
         target_cmpl_cb(rreq);
     }
-
-  fn_exit:
-    return mpi_errno;
-}
-
-static void handle_am_recv(void *request, ucs_status_t status, ucp_tag_recv_info_t * info)
-{
-    return;
 }
 
 int MPIDI_UCX_progress(int vci, int blocking)
 {
     int mpi_errno = MPI_SUCCESS;
     ucp_tag_recv_info_t info;
     MPIDI_UCX_ucp_request_t *ucp_request;
-    void *am_buf;
-    ucp_tag_message_h message_handle;
+
     /* check for active messages */
-    message_handle =
-        ucp_tag_probe_nb(MPIDI_UCX_global.worker, MPIDI_UCX_AM_TAG, MPIDI_UCX_AM_TAG, 1, &info);
-    while (message_handle) {
+    while (true) {
+        ucp_tag_message_h message_handle =
+            ucp_tag_probe_nb(MPIDI_UCX_global.worker, MPIDI_UCX_AM_TAG, MPIDI_UCX_AM_TAG, 1,
+                             &info);
+        if (message_handle == NULL)
+            break;
+
         am_buf = MPL_malloc(info.length, MPL_MEM_BUFFER);
         ucp_request = (MPIDI_UCX_ucp_request_t *) ucp_tag_msg_recv_nb(MPIDI_UCX_global.worker,
                                                                       am_buf,
                                                                       info.length,
                                                                       ucp_dt_make_contig(1),
-                                                                      message_handle,
-                                                                      &handle_am_recv);
+                                                                      message_handle, &am_handler);
         while (!ucp_request_is_completed(ucp_request)) {
```

Comment:
The callback does not happen when this function returns, right (I am assuming that is what `_nb` indicates). So when the callback happen, `am_buf` potentially has been freed, right?

Reply:
OK, I got it. The wait loop ensures the completion. Sorry about my confusion,

Reply:
Right. `while (!ucp_request_is_completed(ucp_request))` blocks until the request is complete, implying the callback has been executed. Then the buffer can be safely freed. Perhaps a comment could be added to explain what is happening.